### PR TITLE
Сервер наносит ответный удар

### PR DESCRIPTION
**Pull request by ilya-cherepanov**
---
:mortar_board: [Сервер наносит ответный удар](https://up.htmlacademy.ru/nodejs-api/1/user/1888439/tasks/8)

### CODE REVIEW

File: src/modules/user/user.entity.ts [hunk 1]
```diff
@@ -1,5 +1,4 @@
-import typegoose, {getModelForClass} from '@typegoose/typegoose';
-import {TimeStamps} from '@typegoose/typegoose/lib/defaultClasses.js';
+import typegoose, {getModelForClass, defaultClasses} from '@typegoose/typegoose';
```
Comment (Delagen) [inline src/modules/user/user.entity.ts:1]: const {prop, modelOptions} = typegoose; можно сюда переместить и убрать дефолтовый импорт

File: src/types/route.interface.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import {NextFunction, Request, Response} from 'express';
+import {HttpMethod} from './http-method.enum.js';
+
+
+export interface RouteInterface {
+  path: string;
+  method: HttpMethod;
+  handler: (req: Request, res: Response, next: NextFunction) => void;
```
Comment (Delagen) [inline src/types/route.interface.ts:8]: handler(this: RouteInterface, req: Request, res: Response, next: NextFunction): void; так наверное будет честнее

Comment (Delagen) [reply src/types/route.interface.ts:8]: ну и сами обработчики как public async loginUser(this:ControllerInterface, _req: Request, _res: Response): Promise<void> { тогда будет всё по честному и контекст в них правильный

**Comment by Delagen** [comment]
Короче я бы предложил такие изменения для корректной типовой системы и сразу ошибку в коде нашел
Index: src/common/controller/controller.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/common/controller/controller.ts b/src/common/controller/controller.ts
--- a/src/common/controller/controller.ts	(revision 3b91eca0cb95987b535b544d6aa3dc23c84934ae)
+++ b/src/common/controller/controller.ts	(date 1652826852319)
@@ -4,7 +4,7 @@
 import {ControllerInterface} from './controller.interface.js';
 import {Response, Router} from 'express';
 import {LoggerInterface} from '../logger/logger.interface.js';
-import { RouteInterface } from '../../types/route.interface.js';
+import {RouteInterface} from '../../types/route.interface.js';
 
 
 @injectable()
@@ -18,7 +18,7 @@
   }
 
   public addRoute(route: RouteInterface): void {
-    this._router[route.method](route.path, asyncHandler(route.handler.bind(this)));
+    this._router[route.method](route.path, asyncHandler((...params) => route.handler.apply(this, params)));
     this.logger.info(`Route registered: ${route.method.toUpperCase()} ${route.path}`);
   }
 
Index: src/types/route.interface.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/types/route.interface.ts b/src/types/route.interface.ts
--- a/src/types/route.interface.ts	(revision 3b91eca0cb95987b535b544d6aa3dc23c84934ae)
+++ b/src/types/route.interface.ts	(date 1652826030286)
@@ -1,9 +1,11 @@
 import {NextFunction, Request, Response} from 'express';
 import {HttpMethod} from './http-method.enum.js';
+import {ControllerInterface} from '../common/controller/controller.interface';
 
 
 export interface RouteInterface {
   path: string;
   method: HttpMethod;
-  handler: (req: Request, res: Response, next: NextFunction) => void;
+
+  handler(this: ControllerInterface, req: Request, res: Response, next: NextFunction):void;
 }
Index: src/cli-command/import-command.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/cli-command/import-command.ts b/src/cli-command/import-command.ts
--- a/src/cli-command/import-command.ts	(revision 3b91eca0cb95987b535b544d6aa3dc23c84934ae)
+++ b/src/cli-command/import-command.ts	(date 1652826700868)
@@ -21,9 +21,6 @@
   private configService!: ConfigInterface;
 
   constructor() {
-    this.onRow = this.onRow.bind(this);
-    this.onEnd = this.onEnd.bind(this);
-
     this.databaseService = new DatabaseService(this.logger);
     this.offerService = new OfferService(OfferModel, this.logger);
     this.configService = new ConfigService(this.logger);
@@ -42,8 +39,8 @@
 
     const fileReader = new TSVFileReader(fileName.trim());
 
-    fileReader.on('row', this.onRow);
-    fileReader.on('end', this.onEnd);
+    fileReader.on('row', (row, resolve) => this.onRow(row, resolve));
+    fileReader.on('end', (count) => this.onEnd(count));
 
     try {
       await fileReader.read();
Index: src/modules/user/user.controller.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/modules/user/user.controller.ts b/src/modules/user/user.controller.ts
--- a/src/modules/user/user.controller.ts	(revision 3b91eca0cb95987b535b544d6aa3dc23c84934ae)
+++ b/src/modules/user/user.controller.ts	(date 1652827571309)
@@ -12,6 +12,7 @@
 import {fillDTO} from '../../utils/other.js';
 import UserDTO from './dto/user.dto.js';
 import HttpError from '../../common/errors/http-error.js';
+import {ControllerInterface} from '../../common/controller/controller.interface';
 
 
 @injectable()
@@ -31,19 +32,19 @@
     this.addRoute({path: '/sign-up', method: HttpMethod.Post, handler: this.createUser});
   }
 
-  public async loginUser(_req: Request, _res: Response): Promise<void> {
+  public async loginUser(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async checkUser(_req: Request, _res: Response): Promise<void> {
+  public async checkUser(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async logoutUser(_req: Request, _res: Response): Promise<void> {
+  public async logoutUser(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async createUser(req: Request<Record<string, unknown>, Record<string, unknown>, CreateUserDTO>, res: Response): Promise<void> {
+  public createUser = async (req: Request<Record<string, unknown>, Record<string, unknown>, CreateUserDTO>, res: Response): Promise<void> => {
     const {body} = req;
 
     const existingUser = await this.userService.findByEmail(body.email);
@@ -54,5 +55,5 @@
 
     const result = await this.userService.create(body, this.config.get('SALT'));
     this.send(res, StatusCodes.CREATED, fillDTO(UserDTO, result));
-  }
+  };
 }
Index: src/app/application.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/app/application.ts b/src/app/application.ts
--- a/src/app/application.ts	(revision 3b91eca0cb95987b535b544d6aa3dc23c84934ae)
+++ b/src/app/application.ts	(date 1652827291180)
@@ -7,6 +7,7 @@
 import {LoggerInterface} from '../common/logger/logger.interface.js';
 import {Component} from '../types/componets.js';
 import {createURI} from '../utils/db.js';
+import {ErrorRequestHandler} from 'express';
 
 
 @injectable()
@@ -34,7 +35,7 @@
   }
 
   private registerExceptionFilters(): void {
-    this.expressApp.use(this.exceptionFilter.catch.bind(this.exceptionFilter));
+    this.expressApp.use((...params: Parameters<ErrorRequestHandler>) => this.exceptionFilter.catch(...params));
   }
 
   public async init(): Promise<void> {
Index: src/modules/offer/offer.controller.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/modules/offer/offer.controller.ts b/src/modules/offer/offer.controller.ts
--- a/src/modules/offer/offer.controller.ts	(revision 3b91eca0cb95987b535b544d6aa3dc23c84934ae)
+++ b/src/modules/offer/offer.controller.ts	(date 1652826877571)
@@ -6,6 +6,7 @@
 import {HttpMethod} from '../../types/http-method.enum.js';
 import {OfferServiceInterface} from './offer-service.interface.js';
 import {Request, Response} from 'express';
+import {ControllerInterface} from "../../common/controller/controller.interface";
 
 
 @injectable()
@@ -28,35 +29,35 @@
     this.addRoute({path: '/favorites/:offerId/:status', method: HttpMethod.Post, handler: this.changeFavoriteStatus});
   }
 
-  public async getOffers(_req: Request, _res: Response): Promise<void> {
+  public async getOffers(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async createOffer(_req: Request, _res: Response): Promise<void> {
+  public async createOffer(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async getOneOffer(_req: Request, _res: Response): Promise<void> {
+  public async getOneOffer(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async updateOffer(_req: Request, _res: Response): Promise<void> {
+  public async updateOffer(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async deleteOffer(_req: Request, _res: Response): Promise<void> {
+  public async deleteOffer(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async getPremiumOffers(_req: Request, _res: Response): Promise<void> {
+  public async getPremiumOffers(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async getFavoriteOffers(_req: Request, _res: Response): Promise<void> {
+  public async getFavoriteOffers(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 
-  public async changeFavoriteStatus(_req: Request, _res: Response): Promise<void> {
+  public async changeFavoriteStatus(this: ControllerInterface, _req: Request, _res: Response): Promise<void> {
     throw new Error('Not implemented!');
   }
 }

File: src/modules/user/user.entity.ts (hunk 1 above)
Comment (ilya-cherepanov) [reply src/modules/user/user.entity.ts:1]: typegoose не полностью поддерживает es модули, поэтому приходится так писать. https://github.com/typegoose/typegoose/issues/172

Comment (Delagen) [reply src/modules/user/user.entity.ts:1]: Ну не надо смотреть на вещи которые были когда то. 12 версия ноды уже даже не на поддержке, 14 истекает уже через год, через полгода LTS будет уже 18. Typescript тоже меняется

Comment (ilya-cherepanov) [reply src/modules/user/user.entity.ts:1]: У меня такая ошибка появляется
error

Comment (Delagen) [reply src/modules/user/user.entity.ts:1]: Всё таки как не крути, а основной формат модулей на текущий момент CommonJS

**Pull request merged**